pageserver/src/tenant/timeline/detach_ancestor.rs

```diff
@@ -116,6 +130,472 @@ impl Default for Options {
+        }
+    }
+}
+
+impl PartialEq<&'_ Attempt> for &'_ ExistingAttempt {
+    fn eq(&self, other: &&Attempt) -> bool {
+        self == *other
+    }
+}
+
+impl PartialEq<&'_ TimelineId> for &'_ ExistingAttempt {
+    fn eq(&self, other: &&TimelineId) -> bool {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => false,
+            ContinuedOverRestart(x) | Actual(x, _) if x == *other => true,
+            _ => false,
+        }
+    }
+}
+
+impl ExistingAttempt {
+    fn start_new(&mut self, detached: &TimelineId) -> Result<completion::Completion, Error> {
+        use ExistingAttempt::*;
+        match self {
+            ReadFromIndexPart => {}
+            Actual(other, barrier) if barrier.is_ready() => {
+                if other != detached {
+                    tracing::warn!(prev=%other, next=%detached, "switching ongoing detach; this is not expected to happen normally, but doesn't necessarily mean anything catastrophic");
+                }
+            }
+            Actual(other, _) | ContinuedOverRestart(other) => {
+                return Err(Error::OtherTimelineDetachOngoing(*other))
+            }
+        }
+
+        let (guard, attempt) = Self::new(detached);
+        *self = attempt;
+        Ok(guard)
+    }
+
+    fn new(detached: &TimelineId) -> (completion::Completion, Self) {
+        let (guard, barrier) = completion::channel();
+        (guard, ExistingAttempt::Actual(*detached, barrier))
+    }
+}
+
+/// Represents an across tenant reset exclusive single attempt to detach ancestor.
+#[derive(Debug)]
+pub(crate) struct Attempt {
+    timeline_id: TimelineId,
+
+    _guard: completion::Completion,
+    gate_entered: Option<utils::sync::gate::GateGuard>,
+}
+
+impl Attempt {
+    pub(crate) fn before_shutdown(&mut self) {
+        let taken = self.gate_entered.take();
+        assert!(taken.is_some());
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct SharedStateBuilder {
+    inprogress: std::collections::HashSet<TimelineId>,
+}
+
+impl SharedStateBuilder {
+    /// While loading, visit a timelines persistent [`crate::tenant::IndexPart`] and record if it is being
+    /// detached.
+    pub(crate) fn record_loading_timeline(
+        &mut self,
+        timeline_id: &TimelineId,
+        index_part: &crate::tenant::IndexPart,
+    ) {
+        if index_part
+            .gc_blocking
+            .as_ref()
+            .is_some_and(|b| b.blocked_by_detach_ancestor())
+        {
+            // if the loading a timeline fails, tenant loading must fail as it does right now, or
+            // something more elaborate needs to be done with this tracking
+            self.inprogress.insert(*timeline_id);
+        }
+    }
+
+    /// Merge the loaded not yet deleting in-progress to the existing datastructure.
+    pub(crate) fn build(self, target: &SharedState) {
+        let mut g = target.inner.lock().unwrap();
+
+        assert_eq!(g.latest.is_none(), g.known_ongoing.is_empty());
```

Comment:
add an assert message, explaining what this is asserting.
I _think_ this is asserting something related to the carrying over of the attempt from before reset, but, not sure.

Reply:
I'll try, the reason is that we assert either the `continue_existing_attempt` happened, or it did not: https://github.com/neondatabase/neon/pull/8430#discussion_r1694852290

Reply:
As it's inlined in #8600, this kind of state assertion no longer exists because we do these in different way around (first scan and set up the gc blocking, then re-establish exclusivity of `Attempt` before upserting the new tenant back into tenantsmap).